File: src/test/java/kitchenpos/menu/tobe/domain/MenuNameTest.java
package kitchenpos.menu.tobe.domain;

import kitchenpos.exception.IllegalNameException;
import kitchenpos.infra.FakePurgomalumClient;
import kitchenpos.infra.PurgomalumClient;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.NullSource;
import org.junit.jupiter.params.provider.ValueSource;

import static org.junit.jupiter.api.Assertions.assertThrows;

public class MenuNameTest {

    PurgomalumClient purgomalumClient;
    MenuNameFactory menuName;

    @BeforeEach
    void setUP() {
        purgomalumClient = new FakePurgomalumClient();
        menuName = new MenuNameFactory(purgomalumClient);
    }

    @ParameterizedTest
    @NullSource
    @DisplayName("메뉴 이름은 필수로 입력해야한다.")
    void nameFail1(final String input) {

        assertThrows(IllegalNameException.class, () -> menuName.create(input));
    }

    @ParameterizedTest
    @ValueSource(strings = {"욕설", "욕설포함"})
    @DisplayName("메뉴 이름에 욕설이 포함되어있는 경우 등록할 수 없다.")
    void nameFail2(final String input) {

        assertThrows(IllegalArgumentException.class, () -> menuName.create(input));
    }
}

File: src/test/kotlin/kitchenpos/menu/tobe/domain/MenuNameTest.kt
package kitchenpos.menu.tobe.domain

import kitchenpos.product.tobe.infra.FakeProfanities
import org.assertj.core.api.Assertions.assertThatIllegalArgumentException
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource
import org.junit.jupiter.params.provider.EmptySource

class MenuNameTest {
    @Test
    @DisplayName("MenuName을 생성한다")
    fun create() {
        val name = "양념치킨"
        val menuName = MenuName(menuNamePolicy = MenuNamePolicy(FakeProfanities()), name = name)

        assertEquals(name, menuName.name)
    }

    @ParameterizedTest
    @DisplayName("MenuName은 필수값이다")
    @EmptySource
    fun createFail(name: String) {
        assertThatIllegalArgumentException().isThrownBy {
            MenuName(menuNamePolicy = MenuNamePolicy(FakeProfanities()), name = name)
        }
    }

    @DisplayName("Profanities의 Profanity가 포함된 MenuName은 생성할 수 없다")
    @ParameterizedTest
    @CsvSource(value = ["욕설메뉴", "비속어"])
    fun createProfanityFail(name: String) {
        assertThatIllegalArgumentException().isThrownBy {
            MenuName(menuNamePolicy = MenuNamePolicy(FakeProfanities(listOf("욕설", "비속어"))), name = name)
        }
    }
}
